## <font color="red">*Exercise 1*</font>

<font color="red">Describe 2 separate predictions relevant to your project and associated texts, which involve predicting text that has not been observed based on patterns that have. Then, in a single, short paragraph, describe a research design through which you could use textual features and the tools of classification and regression to evaluate these predictions.

I have used the data from Indian matrimonial ads for most of my assignments in this course. I have several ads for more than 16 years. The text from these ads could be used to predict ads in the future. 

Moreover, the ads could be used to predict the majority caste that could be found in the newspapers of a particular region.

In order to address either of the questions above, the text from the ads for over 10 years can be used to predict ads in the future. The standard structure for most of these ads is the same owing to constraints of space. Thus, variables based on contects of these ads such as caste, physical attributes etc. can be used to develop a regression model which can capture how much the weight of each oif these in the ads is changing over the years along with support from other topic models or embedding models, and can then be used to predict future ads. 

## <font color="red">*Exercise 2*</font>

<font color="red">Propose a simple causal model in your data, or a different causal model in the annotated Internet Arguments Corpus (e.g., a different treatment, a different outcome), and test it using a linear or logistic regression model. If you are using social media data for your final project, we encourage you to classify or annotate a sample of that data (either compuationally or with human annotators) and examine the effect of texts on replies to that text (e.g., Reddit posts on Reddit comments, Tweets on Twitter replies, YouTube video transcripts on YouTube comments or ratings). You do not need to make a graph of the causal model, but please make it clear (e.g., "X affects Y, and C affects both X and Y.").
    
<font color="red">Also consider using the [ConvoKit datasets](https://convokit.cornell.edu/documentation/datasets.html)! Anytime there is conversation, there is an opportunity to explore the effects of early parts of the conversation on later parts. We will explore this further in Week 8 on Text Generation and Conversation.
    
<font color="red">***Stretch*** (not required): Propose a more robust identification strategy using either matching, difference in difference, regression discontinuity, or an instrumental variable. Each of these methods usually gives you a more precise identification of the causal effect than a unconditional regression. Scott Cunningham's [Causal Inference: The Mixtape](https://mixtape.scunning.com/) is a free textbook on these topics, and all have good YouTube video explanations.

One of the causal questiona that can be explored is "does the level of attack or insult of a post lead to a proportional attack or insult of the comment? In other words, does an attack or insult in the post insinuate the respective feeling in the comments?

In [3]:
#Special module written for this class
#This provides access to data and to helper functions from previous weeks
#Make sure you update it before starting this notebook
import lucem_illud #pip install -U git+git://github.com/UChicago-Computational-Content-Analysis/lucem_illud.git

#All these packages need to be installed from pip
import os #For managing the operating system
import requests #For getting files
import zipfile #For managing zips
import numpy as np #For arrays
import scipy as sp #For some stats
import pandas as pd #Gives us DataFrames
import numpy as np #Math and matrices
import matplotlib.pyplot as plt #For graphics

# statsmodels is a popular Python statistics package
import statsmodels.api as sm
# Let's also import its graphics module
import statsmodels.graphics.api as smg
# And the mediation module
from statsmodels.stats.mediation import Mediation

# Pipelines to add text-based quantiative variables for regressions
! pip install transformers
from transformers import pipeline

# We have a lot of features, so let's set Pandas to show all of them.
pd.set_option('display.max_columns', None)

You should consider upgrading via the '/Users/pranathiiyer/opt/anaconda3/bin/python -m pip install --upgrade pip' command.


In [36]:
url = 'http://nldslab.soe.ucsc.edu/iac/iac_v1.1.zip'

req = requests.get(url)

filename = url.split('/')[-1]
with open(filename,'wb') as output_file:
    output_file.write(req.content)
print('Downloaded file: ' + url)

Downloaded file: http://nldslab.soe.ucsc.edu/iac/iac_v1.1.zip


In [37]:
with zipfile.ZipFile('iac_v1.1.zip') as z:
    with z.open('iac_v1.1/data/fourforums/annotations/mechanical_turk/qr_averages.csv') as f:
        qr = pd.read_csv(f)

    with z.open('iac_v1.1/data/fourforums/annotations/mechanical_turk/qr_meta.csv') as f:
        md = pd.read_csv(f)

# columns = ['key', 'nicenasty', 'questioning-asserting', 'negotiate-attack', 'fact-feeling']
# qr_sub = qr[columns]
# qr_sub = qr

pairs = qr.merge(md, how='inner', on='key')
pairs = pairs[~pairs.quote_post_id.isnull() & ~pairs.response_post_id.isnull()]
pairs

key  discussion_id_x  agree-disagree  agreement  \
0       (731, 1)             6032        0.333333  -1.333333   
1       (660, 3)            10217        0.600000   0.285714   
2       (114, 5)             3462        0.600000  -1.500000   
3        (43, 3)             9930        0.166667  -0.833333   
4      (1314, 0)             5352        0.142857  -1.666667   
...          ...              ...             ...        ...   
9995    (580, 4)              821        0.800000  -2.000000   
9997    (694, 4)             9258        0.000000  -3.200000   
9998    (916, 6)            10301        0.000000  -3.000000   
9999   (1348, 1)             6032        0.857143   0.800000   
10000   (484, 0)             3426        0.142857  -2.833333   

       agreement_unsure    attack  attack_unsure  defeater-undercutter  \
0              0.333333  0.333333       0.000000              0.500000   
1              0.000000  0.714286       0.000000             -2.500000   
2              0.000000  1.333333       0.000000              1.000000   
3              0.333333  1.500000       0.000000              0.400000   
4              0.166667  0.000000       0.166667             -1.166667   
...                 ...       ...            ...                   ...   
9995           0.166667 -0.500000       0.166667              1.000000   
9997           0.200000  0.200000       0.200000             -1.800000   
9998           0.000000 -2.400000       0.000000             -2.400000   
9999           0.600000  0.600000       0.200000              3.000000   
10000          0.166667  0.833333       0.166667             -1.000000   

       defeater-undercutter_unsure  fact-feeling  fact-feeling_unsure  \
0                         0.000000      0.333333             0.333333   
1                         0.000000      1.000000             0.000000   
2                         0.000000      1.500000             0.000000   
3                         0.000000      1.500000             0.166667   
4                         0.166667     -0.833333             0.333333   
...                            ...           ...                  ...   
9995                      0.000000      1.333333             0.166667   
9997                      0.000000      1.600000             0.400000   
9998                      0.000000     -2.400000             0.000000   
9999                      0.000000     -0.600000             0.400000   
10000                     0.000000      0.833333             0.333333   

       negotiate-attack  negotiate-attack_unsure  nicenasty  nicenasty_unsure  \
0              3.000000                 0.250000   0.666667          0.166667   
1             -2.000000                 0.000000   1.142857          0.000000   
2             -1.500000                 0.000000   2.166667          0.000000   
3             -2.000000                 0.000000   1.666667          0.000000   
4              0.833333                 0.333333   0.166667          0.166667   
...                 ...                      ...        ...               ...   
9995          -1.000000                 0.000000   0.500000          0.166667   
9997          -1.200000                 0.200000   0.600000          0.200000   
9998           2.800000                 0.000000  -2.400000          0.000000   
9999          -3.000000                 0.000000   0.600000          0.200000   
10000         -2.000000                 0.000000   1.500000          0.333333   

       personal-audience  personal-audience_unsure  questioning-asserting  \
0              -2.250000                  0.250000              -4.250000   
1              -1.500000                  0.000000               0.500000   
2              -4.000000                  0.000000              -1.500000   
3              -2.800000                  0.000000               0.000000   
4              -3.333333                  0.166667              -0.166667   
...                  ...                   

In [38]:
# Self-merge where the 'response' matches another 'quote' in the DataFrame
triples = pairs.merge(pairs,left_on='response',right_on='quote',how='inner',suffixes=('_r1','_r2'))

# Rename and reorder columns
triples = triples.rename(columns={'quote_r1':'quote', 'quote_r2':'response1', 'response_r2':'response2'})
triples = triples.drop(columns=['response_r1'])
front_columns = [
                 'quote','response1','response2','attack_r1','fact-feeling_r1','nicenasty_r1','sarcasm_r1',
                 'agreement_r2'
                ]
triples = triples.dropna(subset=front_columns)
triples = triples[front_columns].join(triples.drop(columns=front_columns))

# Display triples
triples

quote  \
0     I remember looking at the classic evolutionary...   
1                              What is the fun in that?   
2     First off, the scientific method goes:\n \n 1)...   
3     You can ignore the obvious question. This is w...   
4     Its really sad what these gay predator priests...   
...                                                 ...   
1341                      1. Did she do anything wrong?   
1342  this is also the reason why atheism is more li...   
1343  Assault weapons? first of all, what are assaul...   
1344  And without a gun you're defenseless, why can'...   
1345  What I don't understand is why YEC's want to L...   

                                              response1  \
0     Why do you find it necessary to fit observatio...   
1     Seriously? Well, I come here hoping for someth...   
2     You guys know me. Always happy to correct anyo...   
3     Actually what they are really doing is ignorin...   
4     Homosexuals are attracted to adults of the sam...   
...                                                 ...   
1341  Yes, she walked down the wrong alley, alone. W...   
1342  It is curious how lately christians have been ...   
1343  Thanks Patriot for the well researched respons...   
1344  Hardly any criminals go armed in England, and ...   
1345    That's what faith does. It limits your options.   

                                              response2  attack_r1  \
0     Evolution has no goals, it is merely a beautif...   0.333333   
1     nah, I was just poking fun because I can! Pers...  -0.600000   
2     Ah, thanks for the correction, although there ...   2.400000   
3     Really, then show me how I'm wrong - without d...  -3.500000   
4     Homosexuals are attracted to people of the sam...   0.166667   
...                                                 ...        ...   
1341  So it was wrong for her to walk down a public ...   0.166667   
1342  Well, the type of atheism you are talking abou...   1.200000   
1343  The ban on assault weapons stands on such polt...   2.000000   
1344  All of the British papers keep going on about ...   2.400000   
1345  Here's a nice little Google def: A convinced b...   0.600000   

      fact-feeling_r1  nicenasty_r1  sarcasm_r1  agreement_r2     key_r1  \
0            0.333333      0.666667    0.200000     -2.833333   (731, 1)   
1           -2.200000      0.000000    0.000000     -2.166667   (697, 2)   
2            2.800000      2.200000    0.000000     -0.400000     (9, 0)   
3           -3.166667     -3.166667    0.166667     -1.833333  (1077, 1)   
4            2.166667     -0.166667    0.400000     -2.666667   (611, 0)   
...               ...           ...         ...           ...        ...   
1341         0.166667      0.500000    0.333333     -3.333333    (27, 4)   
1342         0.800000      0.600000    0.250000     -2.250000  (1382, 6)   
1343         2.000000      2.500000    0.000000      0.200000   (897, 4)   
1344         2.000000      2.600000    0.000000     -2.800000  (1231, 1)   
1345        -0.600000      0.600000    0.000000     -2.400000  (1348, 1)   

      discussion_id_x_r1  agree-disagree_r1  agreement_r1  \
0                   6032           0.333333     -1.333333   
1                   5205           0.833333     -2.400000   
2                   9449           0.400000      0.600000   
3                   3467           0.400000     -4.333333   
4                   4337           0.333333     -2.000000   
...                  ...                ...           ...   
1341               13306           0.800000      0.666667   
1342                3982           0.000000     -1.000000   
1343                  50           0.833333      0.500000   
1344               11999           0.285714     -1.600000   
1345                6032           0.857143      0.800000   

      agreement_unsure_r1  attack_unsure_r1  defeater-undercutter_r1  \
0                0.333333          0.000000                 0.500000   
1

In [41]:
y = triples['attack_r2']
X_cols = ['attack_r1']
X = sm.add_constant(triples[X_cols])

lm1 = sm.OLS(y,X).fit()
lm1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              attack_r2   R-squared:                       0.055
Model:                            OLS   Adj. R-squared:                  0.054
Method:                 Least Squares   F-statistic:                     78.03
Date:                Mon, 21 Feb 2022   Prob (F-statistic):           3.12e-18
Time:                        07:35:16   Log-Likelihood:                -2394.8
No. Observations:                1340   AIC:                             4794.
Df Residuals:                    1338   BIC:                             4804.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3582      0.041      8.742      0.000       0.278       0.439
attack_r1      0.2346      0.027      8.834      0.000       0.183       0.287
==============================================================================
Omnibus:                       18.319   Durbin-Watson:                   1.873
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               18.706
Skew:                          -0.286   Prob(JB):                     8.67e-05
Kurtosis:                       3.091   Cond. No.                         1.67
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The p-value for attack_r1 is 0.000, which means that we reject the null hypothesis that there is no effect of attack or insult of the post on that of the comments. However, we have other variables such as the content of he post itself, language used etc. that could act as confounders.

Another causal question that can be explored is "does the factual nature of a comment lead to increased agreement? In other words, do people tend to agree with a post when it feel smore like a fact?

In [45]:
y = triples['agreement_r2']
X_cols = ['fact-feeling_r1']
X = sm.add_constant(triples[X_cols])

lm1 = sm.OLS(y,X).fit()
lm1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           agreement_r2   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     1.866
Date:                Mon, 21 Feb 2022   Prob (F-statistic):              0.172
Time:                        11:19:17   Log-Likelihood:                -2580.2
No. Observations:                1340   AIC:                             5164.
Df Residuals:                    1338   BIC:                             5175.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              -1.2915      0.045    -28.392      0.000      -1.381      -1.202
fact-feeling_r1     0.0415      0.030      1.366      0.172      -0.018       0.101
==============================================================================
Omnibus:                       87.793   Durbin-Watson:                   1.868
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              104.515
Skew:                           0.645   Prob(JB):                     2.02e-23
Kurtosis:                       3.457   Cond. No.                         1.51
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The p-value for fact-feeling_r1 is 0.172, which means that we fail to reject the null hypothesis that there is no effect of fact feeling of the post on agreement. However, this could have confounders. 

## <font color="red">*Exercise 3*</font>

<font color="red">Propose a measure you could generate to fill in or improve upon the simple causal model you proposed above and how you would split the data (e.g., a % of your main data, a separate-but-informative dataset). You do not have to produce the measure.
    
<font color="red">***Stretch*** (not required): Produce the measure and integrate it into your statistical analysis. This could be a great approach for your final project!

One measure of understanding if it is the attacking or insulting nature of the post that resulted in that of the comment could be to ask another set of M turk workers, given that they did not take part in this study, as to how they would respond to posts that are of varying attack and insult levels. This could then be compared to real data to see if a trend emerges. Moreover, they can also be asked why they would respond the way they would. There could be several options such as "language of the post", "context matter" etc. which can eliminate confounding to some extent. Doing this would of course significantly also depend on monetary constraints for the project. 

One measure to check if the fact feeling would result in agreement of the post would be to also ask m-turk workers when they saw a post, if they would agree with it, before showing them the comments to annotate. These results although not perfect, could give more insights into what might affect agreement.

## <font color="red">*Exercise 4*</font>

<font color="red">Propose a mediation model related to the simple causal model you proposed above (ideally on the dataset you're using for your final project). If you have measures for each variable in the model, run the analysis: You can just copy the "Mediation analysis" cell above and replace with your variables. If you do not have measures, do not run the analysis, but be clear as to the effect(s) you would like to estimate and the research design you would use to test them.

There are possibilities of two mediators here. One is the fact-feeling of the first comment which has not been annotated, but which could significantly influence the agreement of the second comment. Moreover, more factual posts might lead to more likes/upvotes, and hence create the notion of collective agreement which is then reflected by a person commenting. The measures for these are not readily available in the dataset and hence, it becomes slightly difficult to run the mediation model. 

## <font color="red">*Exercise 5*</font>

<font color="red">Propose a confounder in your final project data that could be controlled for using the method of Pryzant et al. Generate an appropriate CSV file and run it on some vocabulary in your corpus. If you have no such confounder available for your final project, you can use `complaints.csv` with a new vocabulary or for different variables (e.g., `Sub-product`, `Company`) or you can further explore the interesting case of `mortgage`. Also keep in mind the [ConvoKit datasets](https://convokit.cornell.edu/documentation/datasets.html). In any case, be sure to interpret the results. What does the output of `causal_attribution.score_vocab()` mean in your context?

In [46]:
url = 'https://files.consumerfinance.gov/ccdb/complaints.csv.zip'

req = requests.get(url)

filename = url.split('/')[-1]
with open(filename,'wb') as output_file:
    output_file.write(req.content)
print('Downloaded file: ' + url)

Downloaded file: https://files.consumerfinance.gov/ccdb/complaints.csv.zip


In [47]:
zipfile.ZipFile('complaints.csv.zip').extractall('')

In [48]:
complaints = pd.read_csv('complaints.csv')
complaints

/Users/pranathiiyer/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Date received                                            Product  \
0          2019-05-03                                       Student loan   
1          2022-01-15  Credit reporting, credit repair services, or o...   
2          2021-12-08  Credit reporting, credit repair services, or o...   
3          2021-12-31  Credit reporting, credit repair services, or o...   
4          2022-02-18  Credit reporting, credit repair services, or o...   
...               ...                                                ...   
2493815    2017-02-09                                    Debt collection   
2493816    2015-04-29                                           Mortgage   
2493817    2017-03-31                                   Credit reporting   
2493818    2017-01-16                                   Credit reporting   
2493819    2018-03-07                                           Mortgage   

                         Sub-product  \
0               Private student loan   
1                   Credit reporting   
2                   Credit reporting   
3                   Credit reporting   
4                   Credit reporting   
...                              ...   
2493815                I do not know   
2493816  Conventional fixed mortgage   
2493817                          NaN   
2493818                          NaN   
2493819       Other type of mortgage   

                                                     Issue  \
0                            Struggling to repay your loan   
1                     Incorrect information on your report   
2        Problem with a credit reporting company's inve...   
3        Problem with a credit reporting company's inve...   
4                     Incorrect information on your report   
...                                                    ...   
2493815              Cont'd attempts collect debt not owed   
2493816           Loan modification,collection,foreclosure   
2493817           Credit reporting company's investigation   
2493818             Incorrect information on credit report   
2493819                     Trouble during payment process   

                                                 Sub-issue  \
0        Can't get other flexible options for repaying ...   
1                            Account information incorrect   
2        Their investigation did not fix an error on yo...   
3        Was not notified of investigation status or re...   
4                      Information belongs to someone else   
...                                                    ...   
2493815                  Debt resulted from identity theft   
2493816                                                NaN   
2493817           No notice of investigation status/result   
2493818                                     Account status   
2493819                                                NaN   

                              Consumer complaint narrative  \
0                                                      NaN   
1                                                      NaN   
2        First Progress Card was notified throughout th...   
3        As by Law, under 15 U.S Code 1601- Congression...   
4                                                      NaN   
...                                                    ...   
2493815  I have disputed my debts several times with no...   
2493816  My father died in XX/XX/XXXX. Left me his only...   
2493817  cfbp i would Like to file a complaint on Exper...   
2493818  My husband and I are in the middle of an FHA S...   
2493819                                                NaN   

                                   Company public response  \
0                                                      NaN   
1                                                      NaN   
2                                                      NaN   
3        Company has responded to the consumer and the ...   
4                                                      NaN   
...    

In [49]:
import causal_attribution

In [52]:
complaints['timely'] = complaints['Timely response?'].apply(lambda x: 1 if (x == 'Yes') else 0)

In [50]:
product_num = complaints['Company'].value_counts()
product_num

EQUIFAX, INC.                                  391512
TRANSUNION INTERMEDIATE HOLDINGS, INC.         296132
Experian Information Solutions Inc.            273750
BANK OF AMERICA, NATIONAL ASSOCIATION          109806
WELLS FARGO & COMPANY                           93653
                                                ...  
EMPIRE HOME MORTGAGE, INC.                          1
Rapture Technology                                  1
Zenith Financial & Insurance Solutions Inc.         1
BAY VALLEY                                          1
Beacon Funding Corporation                          1
Name: Company, Length: 6292, dtype: int64

In [53]:
product_timelies = complaints.groupby('Company').agg({'timely': 'sum'}).sort_values(by=['timely'],ascending=False)
pcts = product_timelies.divide(product_num,axis=0).sort_values(by=['timely'],ascending=False)
pcts

timely
Kashable LLC                                 1.0
Evans & Mullinix, PA                         1.0
Equity Settlement Services, Inc.             1.0
SUCCESS MORTGAGE PARTNERS, INC               1.0
Ernst, Ernst & Artmann, Inc.                 1.0
...                                          ...
EBA,LLC                                      0.0
E Z CASH OF DELAWARE, INC.                   0.0
E Mortgage                                   0.0
Downeast Capital, Inc.                       0.0
(Former)Shapiro, Swertfeger & Hasty, LLP     0.0

[6292 rows x 1 columns]

In [58]:
%%time

importance_scores = causal_attribution.score_vocab(
    vocab=[
           # Words I expect to be associated with Product
           'credit','bank','financial','solutions','ameritech','information','billing','collections','company','america',
            'holdings','institution'
           # Words I expect to be associated with timeliness
           'fraud','crime','urgent','please','help',
           # Words I expect to not be associated with either
           'bad','good','help','thank','why','husband','wife','family'
          ],
    csv='complaints.csv',
    delimiter=",",
    name_to_type={
        'Consumer complaint narrative': 'input',
        'Company': 'control',
        'Timely response?': 'predict',
    })

importance_scores['Timely response?']['Yes']

/Users/pranathiiyer/opt/anaconda3/lib/python3.8/site-packages/IPython/core/magic.py:187: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  call = lambda f, *a, **k: f(*a, **k)


CPU times: user 21.9 s, sys: 1.48 s, total: 23.4 s
Wall time: 23.2 s


[('please', 1.5127278695290443),
 ('wife', 0.9561634924029931),
 ('thank', 0.8155689486884512),
 ('UNK', 0.8106419076793827),
 ('bank', 0.7141308058053255),
 ('crime', 0.6699398596538231),
 ('financial', 0.6677981074899435),
 ('husband', 0.5049553451390238),
 ('company', 0.4355579301773105),
 ('collections', 0.37322930939262733),
 ('urgent', 0.3579793081153184),
 ('why', 0.34389245428610593),
 ('credit', 0.3191966889426112),
 ('good', 0.31290884778718464),
 ('help', 0.27814008068526164),
 ('bad', 0.26212206596392207),
 ('family', 0.25406818464398384),
 ('america', 0.19798212311798125),
 ('billing', 0.15621957552502863),
 ('ameritech', 0.07860446604900062),
 ('PAD', 0.025846235803328454),
 ('institutionfraud', -0.017285301524680108),
 ('information', -0.025421714157346287),
 ('holdings', -0.07938882015878335),
 ('help', -0.13497317875771841),
 ('solutions', -0.19022408843738958)]

These results are clearly not very meaningful. They suggest that when controlling for company please is a good way to get a timely response. But several other words such as 'wife', 'crime', 'husband' etc do not make sense. While some words such as 'thank', 'financial' etc do make sense. One of the main reasons for this could be the fact that it is tough to some up with a lexicon for the category of company. This could include words related to finance but also the names of the companies themselves. 

## <font color="red">*Exercise 6*</font>

<font color="red">Pick one other paper on causal inference with text from the ["Papers about Causal Inference and Language
" GitHub repository](https://github.com/causaltext/causal-text-papers). Write at least three sentences summarizing the paper and its logic of design in your own words.
    
<font color="red">***Stretch*** (not required): Skim a few more papers. The causal world is your textual oyster!

https://ojs.aaai.org/index.php/AAAI/article/view/9205/9064 --> I read this paper!

The authors try to estimate the effects of exercise on mental health by using tweets that express anxiety, depression, and anger. They use an approach of matching by utilizing information related to gender from US census data, geo location, and other information available to tackle confounding bias. Their observational study was successful in finding that there are lesser incidences of tweets expressing anxiety and depression amongst ussers engaged in physical activity, which was captured using their use of physical activity trackers.